Kontrol documentation and cleanup

mds1 opened the pull request:
**Description**

Various improvements to Kontrol docs and implementation:
- doc: Updates the README with more details.
- ci: Check that kontrol state diffs are up to date as part of snapshot checks.
    - This required refactoring the kontrol shell scripts to reuse some of the same docker logic, since in CI we run kontrol via docker. This shared logic lives in `common.sh`
- perf: Improve proof assumptions and better document them (need to see how this affects kontrol run times in CI before merging).
- feat: cleanup `make-deployment-summary.sh` even if forge script fails.
- chores: fix typos, update `run-kontrol.sh` help text, simplify `kprove` foundry profile, remove unnecessary braces on var names

**Tests**

No tests are added, but existing tests are modified.

I tested locally to verify the updated snapshot check will exit with an error if the `snapshots/state-diff/Kontrol-Deploy.json` file is not up to date.

**Additional context**

Before merging, I'd like to also get a review from @JuanCoRo to verify everything is accurate

mds1 committed:
docs: change directory structure format

mds1 committed:
build: add kontrol installation commands

mds1 committed:
chore: simpler kprove profile

mds1 committed:
chore: update help text in run-kontrol.sh

mds1 committed:
docs: refactor and clarify README content

mds1 committed:
chore: fix typo in test name

mds1 committed:
ci: check kontrol state diff in snapshot checks

mds1 committed:
feat: cleanup even when forge script fails

mds1 committed:
perf: improve assumptions and better document them

mds1 committed:
doc: more docs

JuanCoRo commented on packages/contracts-bedrock/test/kontrol/README.md:
```diff
@@ -1,98 +1,129 @@
-- [`interfaces`](./proofs/interfaces): Files with the signature of the functions involved in the verification effort
-- [`utils`](./proofs/utils): Proof dependencies, including the summary contracts
+1. `cd` to the root of this repo.
+2. Install Foundry by running `pnpm install:foundry`. This installs `foundryup`, the foundry toolchain installer, then installs the required foundry version.
+3. Install Kontrol by running `pnpm install:kontrol`. This installs `kup`, the package manager for RV tools, then installs the required kontrol version.
```
Would it be good to add at the end something like 
"Alternatively, run `curl -L https://foundry.paradigm.xyz | bash`" for Foudry and 
"Alternatively, run `bash <(curl https://kframework.org/install)` and `kup install kontrol --version v$(jq -r .kontrol < versions.json)`" for Kontrol? Perhaps not, but I thought I'd ask

mds1 replied:
Going to leave this as-is :)

JuanCoRo commented on packages/contracts-bedrock/test/kontrol/README.md:
````diff
@@ -1,98 +1,129 @@
-```bash
-  ./test/kontrol/scripts/run-kontrol.sh $option
+[`KontrolDeployment.sol`](./deployment/KontrolDeployment.sol) contains the minimal deployment sequence required by the proofs.
+The [`make-summary-deployment.sh`](./scripts/make-summary-deployment.sh) script will generate a summary contract is used to load the post-`setUp` state directly into Kontrol.
````
```suggestion
The [`make-summary-deployment.sh`](./scripts/make-summary-deployment.sh) script will generate a JSON state diff. This state diff is used in two ways by Kontrol.
First, Kontrol generates a summary contract recreating the state diffs recorded in the JSON. This contract is used to test that the information contained in the generated JSON is correct and aids in the specification of the symbolic property tests. The generation of this contract is also handled by the `make-summary-deployment.sh` script.
Second, the state diff JSON is used to load the post-deployment state directly into Kontrol when running the proofs.
```

mds1 committed:
doc: update description of make-summary-deployment.sh

Co-authored-by: Juan C. <38925412+JuanCoRo@users.noreply.github.com>

JuanCoRo commented on packages/contracts-bedrock/test/kontrol/README.md:
```diff
@@ -1,98 +1,129 @@
+[`KontrolDeployment.sol`](./deployment/KontrolDeployment.sol) contains the minimal deployment sequence required by the proofs.
+The [`make-summary-deployment.sh`](./scripts/make-summary-deployment.sh) script will generate a summary contract is used to load the post-`setUp` state directly into Kontrol.
+
+This step is optional if an up-to-date summary contract already exists, which will be the case until the `KontrolDeployment` contract changes.
```
If any of the tested contracts changes, it should also be the case that the JSON state diff (and thence the deployment contract) changes, since the etched code will be different

mds1 replied:
Clarified in 1d24faf45

mds1 committed:
perf: revert back to the 320 byte assumption for now

mds1 committed:
doc: clarify optionality of build step

mds1 committed:
fix: add output directory to shellcheck ignored dirs

mds1 committed:
fix/refactor: snapshots job needs kontrol via docker in CI

mds1 committed:
fix: shellcheck lints

Inphi commented on packages/contracts-bedrock/test/kontrol/README.md:
```diff
@@ -1,98 +1,131 @@
+### Build Deployment Summary
 
-The verification execution consists of two steps, although the first step may be omitted to use the committed version. There's one script to run per step. These scripts should be run from the [`contracts-bedrock`](../../) directory.
+First, generate a deployment summary contract from the deploy script in [`KontrolDeployment.sol`](./KontrolDeployment.sol) by running the following command:
```
the KontrolDeployment.sol link target is broken here and elsewhere

mds1 replied:
Thanks, fixed in 1861fde98

Inphi commented on packages/contracts-bedrock/test/kontrol/scripts/run-kontrol.sh:
```diff
@@ -215,7 +119,7 @@ regen=
 test_list=( "OptimismPortalKontrol.prove_finalizeWithdrawalTransaction_paused" \
```
might be useful to document when to update this list in the README.

mds1 replied:
Great point, added a brief "Add New Proofs" section in 1861fde98 which explains this. I plan to write more details here in a future PR

Inphi reviewed (approved):
LGTM!

mds1 committed:
docs: fix links and add brief summary of adding new proofs

mds1 committed:
fix: remove shift, which cause early exit

we only have 1 input param so don't need shift: when shift fails
it would cause the script to exit early due to set -e

mds1 committed:
fix: run in proper environment

mds1 committed:
fix: start docker when needed

mds1 committed:
chore: cleanup docker

mds1 committed:
fix: another fix from shell script refactor

mds1 committed:
ci: add docker to bedrock tests

trianglesphere commented on .circleci/config.yml:
```diff
@@ -365,6 +365,8 @@ jobs:
       - attach_workspace: { at: "." }
       - check-changed:
           patterns: contracts-bedrock,op-node
+      - setup_remote_docker:
```
This doesn't seem to be enough to set up docker for the kontrol job

mds1 replied:
Yea trying to figure out why. It seems to be sufficient for the `kontrol-tests` job but not the `contracts-bedrock-tests` job, but the "Setup a remote Docker engine" step is still not present for the bedrock tests . Maybe related to `- attach_workspace: { at: "." }`, but still investigating...

<img width="1516" alt="image" src="https://github.com/ethereum-optimism/optimism/assets/17163988/8691b04b-f5c1-4629-a28f-770e118b3ad5">

trianglesphere replied:
@mds1 check the underly executor (i.e. Docker vs remote docker)

mds1 replied:
I had `setup_remote_docker` under the `contracts-bedrock-tests` job, not the `contracts-bedrock-checks` job 🤦

Pushed 4d4135fda which should fix it

trianglesphere replied:
yep, docker is running. Next error is an invalid symlink in typescript land

mds1 replied:
Investigating—I can reproduce this locally by running the `make-deployment-summary.sh` and `run-kontrol.sh` scripts in container mode

mds1 replied:
Ok got it all passing with a few fixes, so resolving this thread

mds1 committed:
fix(ci): move remote docker from contracts-bedrock-tests to contracts-bedrock-checks

mds1 committed:
ci: maybe docker setup needs to come earlier

mds1 committed:
fix: various changes so make-summary-deployment works in docker

- avoid copying node_modules to container, which caused a symlink error
- fix which script commands are run in docker
- restore order of setup_remote_docker step

mds1 committed:
fix: bind mount instead of copy into docker

mds1 committed:
style: remove unneded braces

mds1 commented on packages/contracts-bedrock/test/kontrol/scripts/common.sh:
None of the code in this file is new, just refactored out of `run-kontro.sh` so it can be reused in `make-deployment-summary.sh`. The only change is using a bind mount in the `docker run` command instead of copying files over after. This was done so files generated by `make-deployment-summary.sh` are available in the docker container

mds1 committed:
fix: -v to create dir on docker, restore file copying

mds1 committed:
fix: copy new files

mds1 committed:
fix permissions error
